### DISCUSSION TITLE
Setup fails - domain validation fails with my own CA

### DISCUSSION THREAD

**Discussion by treblada** [nextcloud/all-in-one#2962] [Questions] [answered]
I try to setup Nextcloud with AIO and my own CA + server certificate. Reverse proxy is configured and running correctly.

### Steps to reproduce
1. Configure your own local TLD in your DNS resolver (e.g. .mine)
2. Configure the reverse proxy as described in the documentation
3. Set NEXTCLOUD_TRUSTED_CACERTS_DIR in the initial master container command
4. put a CA cert PEM file in  NEXTCLOUD_TRUSTED_CACERTS_DIR
5. start the master container with

`docker run \
	--detach \
	--env NEXTCLOUD_TRUSTED_CACERTS_DIR=$HOME/nextcloud/ca \
	--env NEXTCLOUD_DATADIR=/srv/nextcloud \
	--env NEXTCLOUD_MOUNT=/srv \
	--env APACHE_PORT=11000 \
	--env APACHE_IP_BINDING=127.0.0.1 \
	--sig-proxy=false \
	--name nextcloud-aio-mastercontainer \
	--restart always \
	--publish 8080:8080 \
	--volume nextcloud_aio_mastercontainer:/mnt/docker-aio-config \
	--volume /var/run/docker.sock:/var/run/docker.sock:ro \
	nextcloud/all-in-one:latest
`

### Expected behavior 
Domain validation succeeds, the server certificate is validated by the master container using the provided additional CA

### Actual behavior 
Mastercontainer refuses to accept the cert. An error message says
`Domain does not point to this server or the reverse proxy is not configured correctly. See the mastercontainer logs for more details. ('sudo docker logs -f nextcloud-aio-mastercontainer')`

And the docker log says

`NOTICE: PHP message: The response of the connection attempt to "https://nc.mine:443" was: 
NOTICE: PHP message: Expected was: e8ac75846f1d23dead004bd44c27f7880565a493d3b575c0
NOTICE: PHP message: The error message was: SSL certificate problem: unable to get local issuer certificate`

I've tried to search for the certificate name in the nextcloud-aio-mastercontainer and nexctcloud-aio-domaincheck containers, but I could not find it. My assumption was that I would find my CA cert somewhere in the container.

`docker exec -it nexctcloud-aio-mastercontainer find / -name "mycacertname"`

### Host OS 
Linux lion 5.15.0-76-generic #83-Ubuntu SMP Thu Jun 15 19:16:32 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

#### Nextcloud AIO version 
Nextcloud AIO v6.2.1

#### Current channel 
I do not see any channel name in the setup view.

#### Other valuable info 
My CA and the certificate in the reverse proxy are working fine. I can access the future nextcloud server directly at https://nc.mine:443 and I do receive the expected hash. I have imported my own CA in my browser before.
Yet the mastercontainer seems not to be able to verify the certificate.

**Comment by treblada** [comment]
Update: I tried to ignore this problem with `SKIP_DOMAIN_VALIDATION=true`, but the installation just runs into trouble later, the nextcloud-aio-nextcloud container cannot start because of a certificate which cannot be validated.

**Comment by szaimen** [comment, marked as answer]
Hi, see https://github.com/nextcloud/all-in-one#are-self-signed-certificates-supported-for-nextcloud and https://github.com/nextcloud/all-in-one/discussions/588#discussioncomment-3450343

**Reply by treblada** [reply]
Hi, thank thank you for taking the time to answer. Unfortunately, that does not match my case at all. I do NOT use a self signed certificate. I have my own CA which signs a certificate used in the reverse proxy. I followed the local installation instructions to the point where they failed me. 
It is my understanding that I can provide any number of additional CAs to the master container via NEXTCLOUD_TRUSTED_CACERTS_DIR env variable, but that seems not to be working.
It looks very much like a bug to me, but I might be mistaken.

Can you please tell me, where the additional CAs are mounted in the mastercontainer? I would like to verify that they exist in the the context of the container.

**Reply by Zoey2936** [reply]
they are only mounted into the nextcloud container if I remember correctly. And again self signed certificates aren't supported

**Reply by treblada** [reply]
Thank you @Zoey2936, I will have a look. But once again: it is not a self-signed certificate. It is a certificate signed by a CA which is not included in the default container. That is what NEXTCLOUD_TRUSTED_CACERTS_DIR exists for, right? To add CAs which are not included in the default.

**Reply by szaimen** [reply]
> That is what NEXTCLOUD_TRUSTED_CACERTS_DIR exists for, right? To add CAs which are not included in the default.

No, this one is only for the nextcloud container as the name of the variable  states. It is mostly there to make ldaps work if ldaps should be using a self-singed certificate.

**Reply by kebugcheckex** [reply]
What are the considerations leading to not supporting self-signed certificates or private CAs?

I want to deploy nextcloud in the internal network of an organization and this organization has its own CA signing certificates for internal servers. All devices issued to employees are preconfigured so that they trust the private root CA.

**Comment by kspanak** [comment]
A workaround that worked for me:
* start the AIO container and then open container's cli: `sudo docker exec -it nextcloud-aio-mastercontainer /bin/bash`
* get the contents of _ca-certificates.crt_: `cat /etc/ssl/certs/ca-certificates.crt`
* copy-paste the above contents and also append the contents of your _customCA.crt_ to a new file: `ca-certificates-with-appended-customCA.crt` (in host's file system)
* stop the AIO container and clear everything (**Be careful** what you delete here! I haven't tested without clearing first.)
    ```bash
    sudo docker ps -a
    sudo docker rm # add-here-all-the-ids-of-nextcloud-containers

    sudo docker volume ls
    sudo docker volume rm # add-here-all-the-nextcloud-volume-names

    # if NEXTCLOUD_DATADIR is used, delete the volume or the contents of the directory to which it points
    ```
* edit the `docker-compose.yml` or _docker command_ by **adding** in the volumes section the equivalent of:
    ```bash
        volumes:
          - /path/to/ca-certificates-with-appended-customCA.crt:/etc/ssl/certs/ca-certificates.crt:ro
    ```

I hope that this helps somebody.
